presto-main/src/main/java/com/facebook/presto/sql/planner/Plan.java

```diff
@@ -13,22 +13,53 @@
  */
 package com.facebook.presto.sql.planner;
 
+import com.facebook.presto.Session;
+import com.facebook.presto.cost.CachingCostProvider;
+import com.facebook.presto.cost.CachingStatsProvider;
+import com.facebook.presto.cost.CostCalculator;
+import com.facebook.presto.cost.CostProvider;
+import com.facebook.presto.cost.PlanNodeCostEstimate;
+import com.facebook.presto.cost.PlanNodeStatsEstimate;
+import com.facebook.presto.cost.StatsCalculator;
+import com.facebook.presto.cost.StatsProvider;
 import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanNodeId;
+import com.facebook.presto.sql.planner.plan.PlanVisitor;
+import com.google.common.collect.ImmutableMap;
 
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.sql.planner.iterative.Lookup.noLookup;
 import static java.util.Objects.requireNonNull;
 
 public class Plan
 {
     private final PlanNode root;
     private final TypeProvider types;
+    private final Map<PlanNodeId, PlanNodeStatsEstimate> stats;
+    private final Map<PlanNodeId, PlanNodeCostEstimate> costs;
 
-    public Plan(PlanNode root, TypeProvider types)
+    public Plan(PlanNode root, TypeProvider types, Map<PlanNodeId, PlanNodeStatsEstimate> stats, Map<PlanNodeId, PlanNodeCostEstimate> costs)
     {
         requireNonNull(root, "root is null");
         requireNonNull(types, "types is null");
+        requireNonNull(stats, "stats is null");
+        requireNonNull(costs, "costs is null");
 
         this.root = root;
         this.types = types;
+        this.stats = stats;
+        this.costs = costs;
+    }
+
+    public static Plan createPlan(PlanNode root, TypeProvider types, StatsCalculator statsCalculator, CostCalculator costCalculator, Session session)
+    {
+        StatsProvider statsProvider = new CachingStatsProvider(statsCalculator, session, types);
+        CostProvider costProvider = new CachingCostProvider(costCalculator, statsProvider, Optional.empty(), noLookup(), session, types);
+        StatsAndCostsComputer statsAndCostsComputer = new StatsAndCostsComputer(statsProvider, costProvider);
+        root.accept(statsAndCostsComputer, null);
+        return new Plan(root, types, statsAndCostsComputer.statsBuilder.build(), statsAndCostsComputer.costsBuilder.build());
```

Comment:
the (`Plan`'s) constructor is not the best place to do this, especially since calculating stats may involve a lot, including network communication